File: api/schemas/rules.py
from typing import Dict, List, Literal, Optional
from pydantic import BaseModel, ConfigDict, Field, model_validator

from core.metrics import (
    RULE_MAX_SCOPE_IDS,
    normalize_rule_scope,
    validate_runtime_rule,
)


RuleGroupIcon = Literal["backlog", "shield", "rocket", "flask", "custom"]


class RuleScopeItem(BaseModel):
    """Which entities an attached rule may act on.

    Scope only narrows the search; the level a rule acts on is its own `level`.
    "account" keeps the historical behaviour of sweeping the whole ad account.
    """

    model_config = ConfigDict(extra="forbid")

    level: Literal["account", "campaign", "adset"] = "account"
    ids: List[str] = Field(default_factory=list, max_length=RULE_MAX_SCOPE_IDS)

    @model_validator(mode="after")
    def validate_scope_shape(self):
        # One source of truth: the runtime validator the worker also uses.
        normalize_rule_scope(self.model_dump())
        return self


class ConditionItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    metric: Literal["spend", "cpl", "cpreg", "cpp", "leads", "registrations", "purchases", "ctr", "cpc"] = "spend"
    operator: Literal["gte", "gt", "lte", "lt", "eq"] = "gte"
    value: float = Field(default=0.0, ge=0, le=1_000_000_000, allow_inf_nan=False)
    time_window: Literal["today", "yesterday", "last_3d", "last_7d"] = "today"


class RulePresetItem(BaseModel):
    id: int
    name: str
    action: str
    # Where the rule reads metrics and applies its action.
    level: Literal["campaign", "adset", "ad"] = "adset"
    enabled: bool = True
    conditions: List[ConditionItem]
    condition_logic: str = "and"
    cooldown_minutes: int = 0
    check_interval_minutes: int = 5
    budget_change_percent: float = 0.0
    budget_max_daily: float = 0.0
    currency_mode: Literal["account"] = "account"
    created_at: str
    # Legacy or unsafe snapshots are held back from execution until re-saved.
    needs_review: bool = False
    review_reason: str = ""
    # Last RULE_ACTION audit event for this preset as ISO 8601 with offset;
    # empty when the rule has never fired.
    last_run_at: str = ""
    # Ad accounts this preset is currently attached to, so the UI can tell a
    # configured rule apart from one that cannot run anywhere yet.
    attached_account_ids: List[str] = Field(default_factory=list)
    # Scope per attached account, so a client can say what detaching removes
    # instead of silently discarding a rule narrowed to specific campaigns.
    attached_scopes: Dict[str, RuleScopeItem] = Field(default_factory=dict)


class CreatePresetRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=120)
    action: Literal["turn_off", "notify_only", "turn_on", "increase_budget", "decrease_budget"] = "turn_off"
    level: Literal["campaign", "adset", "ad"] = "adset"
    enabled: bool = True
    conditions: List[ConditionItem] = Field(min_length=1, max_length=20)
    condition_logic: Literal["and", "or"] = "and"
    cooldown_minutes: int = Field(default=0, ge=0, le=10_080)
    check_interval_minutes: int = Field(default=5, ge=1, le=1_440)
    budget_change_percent: float = Field(default=0.0, ge=0, le=100, allow_inf_nan=False)
    budget_max_daily: float = Field(default=0.0, ge=0, le=10_000_000, allow_inf_nan=False)

    @model_validator(mode="after")
    def validate_safe_action_parameters(self):
        validate_runtime_rule(
            {
                "action": self.action,
                "level": self.level,
                "conditions": [condition.model_dump() for condition in self.conditions],
                "logic": self.condition_logic,
                "cooldown_minutes": self.cooldown_minutes,
                "check_interval": self.check_interval_minutes,
                "budget_change_percent": self.budget_change_percent,
                "budget_max_daily": self.budget_max_daily,
            }
        )
        return self


class RuleGroupWriteRequest(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str = Field(default="", max_length=500)
    icon: RuleGroupIcon = "custom"
    position: Optional[int] = None
    preset_ids: List[int] = Field(default_factory=list, min_length=0, max_length=50)


class RuleGroupResponse(BaseModel):
    id: int
    name: str
    description: str
    icon: str = "custom"
    position: int = 0
    preset_ids: List[int]
    rules: List[RulePresetItem]
    created_at: str


class RuleGroupsReorderRequest(BaseModel):
    group_ids: List[int] = Field(min_length=0, max_length=100)


class ApplyPresetRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    preset_id: int = Field(gt=0)
    scope: RuleScopeItem = Field(default_factory=RuleScopeItem)


class DeletedItemResponse(BaseModel):
    id: int
    kind: Literal["rule", "rule_group"]
    entity_id: int
    name: str
    deleted_at: str
    # Past this moment the item is gone for good.
    purge_at: str
    deleted_by: str = ""


class RestoreDeletedItemResponse(BaseModel):
    kind: Literal["rule", "rule_group"]
    entity_id: int
    name: str
    # Ad accounts a restored rule could not return to: gone, or now running a
    # rule it would contradict.
    skipped_account_ids: List[str] = Field(default_factory=list)
    # Rules of a restored group that were deleted in the meantime.
    missing_rule_ids: List[int] = Field(default_factory=list)

File: frontend/src/lib/rules.ts
import { apiRequest } from '@/lib/api';

/**
 * Wire format of `api/routers/rules.py`. Field names mirror the Pydantic
 * schemas exactly; translation into view models happens in this module so no
 * component has to know the backend shape.
 */

export type RuleMetric =
  | 'spend'
  | 'cpl'
  | 'cpreg'
  | 'cpp'
  | 'leads'
  | 'registrations'
  | 'purchases'
  | 'ctr'
  | 'cpc';

export type RuleOperator = 'gte' | 'gt' | 'lte' | 'lt' | 'eq';

export type RuleTimeWindow = 'today' | 'yesterday' | 'last_3d' | 'last_7d';

export type RuleAction =
  | 'turn_off'
  | 'notify_only'
  | 'turn_on'
  | 'increase_budget'
  | 'decrease_budget';

export type RuleExecutionLevel = 'campaign' | 'adset';

export type RuleGroupIcon = 'backlog' | 'shield' | 'rocket' | 'flask' | 'custom';

/**
 * Which ad sets an attached rule may act on. A rule always acts on ad sets;
 * scope only narrows the search. "account" is the historical behaviour of
 * sweeping every ad set in the ad account.
 */
export type RuleScopeLevel = 'account' | 'campaign' | 'adset';

export interface RuleScope {
  level: RuleScopeLevel;
  ids: string[];
}

/** One rule as stored on an ad account, in the runtime snapshot format. */
export interface AttachedRule {
  preset_id: number;
  name: string;
  scope?: RuleScope;
}

export const ACCOUNT_SCOPE: RuleScope = { level: 'account', ids: [] };

export function attachedRuleScope(rule: AttachedRule): RuleScope {
  return rule.scope ?? ACCOUNT_SCOPE;
}

export interface RuleConditionPayload {
  metric: RuleMetric;
  operator: RuleOperator;
  value: number;
  time_window: RuleTimeWindow;
}

export interface RulePresetPayload {
  id: number;
  name: string;
  action: RuleAction;
  level: RuleExecutionLevel;
  enabled: boolean;
  conditions: RuleConditionPayload[];
  condition_logic: 'and' | 'or';
  cooldown_minutes: number;
  check_interval_minutes: number;
  notify_tg: boolean;
  budget_change_percent: number;
  budget_max_daily: number;
  currency_mode: 'account';
  created_at: string;
  needs_review: boolean;
  review_reason: string;
  last_run_at: string;
  attached_account_ids: string[];
}

export interface RuleGroupPayload {
  id: number;
  name: string;
  description: string;
  icon: RuleGroupIcon;
  position: number;
  preset_ids: number[];
  rules: RulePresetPayload[];
  created_at: string;
}

export interface RulePresetWriteRequest {
  name: string;
  action: RuleAction;
  level: RuleExecutionLevel;
  enabled: boolean;
  conditions: RuleConditionPayload[];
  condition_logic: 'and' | 'or';
  cooldown_minutes: number;
  check_interval_minutes: number;
  notify_tg: boolean;
  budget_change_percent: number;
  budget_max_daily: number;
}

export interface RuleGroupWriteRequest {
  name: string;
  description: string;
  icon: RuleGroupIcon;
  position?: number;
  preset_ids: number[];
}

/* ---------------------------------------------------------------- labels -- */

export const RULE_METRIC_LABELS: Record<RuleMetric, string> = {
  spend: 'Spend',
  cpl: 'CPL',
  cpreg: 'CPReg',
  cpp: 'CPP',
  leads: 'Leads',
  registrations: 'Registrations',
  purchases: 'Purchases',
  ctr: 'CTR',
  cpc: 'CPC',
};

export const RULE_OPERATOR_LABELS: Record<RuleOperator, string> = {
  gte: '≥',
  gt: '>',
  lte: '≤',
  lt: '<',
  eq: '=',
};

export const RULE_TIME_WINDOW_LABELS: Record<RuleTimeWindow, string> = {
  today: 'today',
  yesterday: 'yesterday',
  last_3d: 'last 3d',
  last_7d: 'last 7d',
};

export const RULE_LEVEL_LABELS: Record<RuleExecutionLevel, string> = {
  adset: 'Ad set',
  campaign: 'Campaign',
};

export const RULE_ACTION_LABELS: Record<RuleAction, string> = {
  turn_off: 'TURN OFF',
  turn_on: 'TURN ON',
  notify_only: 'NOTIFY',
  increase_budget: 'BUDGET +',
  decrease_budget: 'BUDGET −',
};

const PERCENT_METRICS: ReadonlySet<RuleMetric> = new Set(['ctr']);

/**
 * Values are stored in each ad account's own currency, so a rule that spans
 * accounts has no single symbol to render. The metric name carries the unit.
 */
function formatConditionValue(metric: RuleMetric, value: number): string {
  const rounded = Number.isInteger(value) ? String(value) : value.toFixed(2);
  if (PERCENT_METRICS.has(metric)) return `${rounded}%`;
  return rounded;
}

export function formatCondition(preset: RulePresetPayload): string {
  if (preset.conditions.length === 0) return 'No conditions';
  const joiner = preset.condition_logic === 'or' ? ' OR ' : ' & ';
  const parts = preset.conditions.map((condition) => {
    const metric = RULE_METRIC_LABELS[condition.metric] ?? condition.metric;
    const operator = RULE_OPERATOR_LABELS[condition.operator] ?? condition.operator;
    return `${metric} ${operator} ${formatConditionValue(condition.metric, condition.value)}`;
  });
  return `IF ${parts.join(joiner)}`;
}

/**
 * Includes the level, because "turn off" means very different things when it
 * takes down one ad set versus a whole campaign.
 */
export function formatAction(preset: RulePresetPayload): string {
  if (preset.action === 'increase_budget') {
    return `BUDGET +${preset.budget_change_percent}%`;
  }
  if (preset.action === 'decrease_budget') {
    return `BUDGET −${preset.budget_change_percent}%`;
  }
  const label = RULE_ACTION_LABELS[preset.action] ?? preset.action.toUpperCase();
  if (preset.action === 'notify_only') return label;
  return `${label} ${preset.level === 'campaign' ? 'CAMPAIGN' : 'AD SET'}`;
}

/**
 * Semantic tone of an action, so a badge never infers meaning from its own
 * label text. "stop" takes delivery away, "positive" adds budget or delivery.
 */
export type RuleActionTone = 'stop' | 'positive' | 'default';

export function ruleActionTone(action: RuleAction): RuleActionTone {
  if (action === 'turn_off' || action === 'decrease_budget') return 'stop';
  if (action === 'turn_on' || action === 'increase_budget') return 'positive';
  return 'default';
}

const MINUTE = 60_000;
const HOUR = 60 * MINUTE;
const DAY = 24 * HOUR;

/** Relative time for the Last run column; empty input means the rule never fired. */
export function formatRelativeTime(isoTimestamp: string, now: number = Date.now()): string {
  if (!isoTimestamp) return 'Never';
  const parsed = Date.parse(isoTimestamp);
  if (Number.isNaN(parsed)) return 'Never';
  const elapsed = now - parsed;
  if (elapsed < MINUTE) return 'Just now';
  if (elapsed < HOUR) return `${Math.floor(elapsed / MINUTE)}m ago`;
  if (elapsed < DAY) return `${Math.floor(elapsed / HOUR)}h ago`;
  return `${Math.floor(elapsed / DAY)}d ago`;
}

/**
 * Ad accounts arrive as Meta ids (`act_123…`). The list only needs to say
 * whether the rule can run at all and in how many places.
 */
export function formatScope(preset: RulePresetPayload): string {
  const count = preset.attached_account_ids.length;
  if (count === 0) return 'Not attached';
  return count === 1 ? '1 ad account' : `${count} ad accounts`;
}

/* ------------------------------------------------------------- requests -- */

export function fetchRulePresets(): Promise<RulePresetPayload[]> {
  return apiRequest<RulePresetPayload[]>('/api/presets');
}

export function fetchRuleGroups(): Promise<RuleGroupPayload[]> {
  return apiRequest<RuleGroupPayload[]>('/api/rule-groups');
}

export function createRulePreset(
  payload: RulePresetWriteRequest,
): Promise<RulePresetPayload> {
  return apiRequest<RulePresetPayload>('/api/presets', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export function updateRulePreset(
  presetId: number,
  payload: RulePresetWriteRequest,
): Promise<RulePresetPayload> {
  return apiRequest<RulePresetPayload>(`/api/presets/${presetId}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  });
}

export function deleteRulePreset(presetId: number): Promise<{ success: boolean }> {
  return apiRequest<{ success: boolean }>(`/api/presets/${presetId}`, {
    method: 'DELETE',
  });
}

export function createRuleGroup(
  payload: RuleGroupWriteRequest,
): Promise<RuleGroupPayload> {
  return apiRequest<RuleGroupPayload>('/api/rule-groups', {
    method: 'POST',
    body: JSON.stringify(payload),
  });
}

export function updateRuleGroup(
  groupId: number,
  payload: RuleGroupWriteRequest,
): Promise<RuleGroupPayload> {
  return apiRequest<RuleGroupPayload>(`/api/rule-groups/${groupId}`, {
    method: 'PUT',
    body: JSON.stringify(payload),
  });
}

export function deleteRuleGroup(groupId: number): Promise<{ success: boolean }> {
  return apiRequest<{ success: boolean }>(`/api/rule-groups/${groupId}`, {
    method: 'DELETE',
  });
}

/* ------------------------------------------- rules attached to accounts -- */

interface AccountRulesResponse {
  account_id: string;
  active_rules: AttachedRule[];
  rules_enabled: boolean;
}

export function assignRuleToAccount(
  accountId: string,
  presetId: number,
  scope: RuleScope,
): Promise<AccountRulesResponse> {
  return apiRequest<AccountRulesResponse>(
    `/api/accounts/${encodeURIComponent(accountId)}/assign-rule`,
    {
      method: 'POST',
      body: JSON.stringify({ preset_id: presetId, scope }),
    },
  );
}

export function setAttachedRuleScope(
  accountId: string,
  presetId: number,
  scope: RuleScope,
): Promise<AccountRulesResponse> {
  return apiRequest<AccountRulesResponse>(
    `/api/accounts/${encodeURIComponent(accountId)}/rules/${presetId}/scope`,
    {
      method: 'PUT',
      body: JSON.stringify(scope),
    },
  );
}

export function detachRuleFromAccount(
  accountId: string,
  presetId: number,
): Promise<AccountRulesResponse> {
  return apiRequest<AccountRulesResponse>(
    `/api/accounts/${encodeURIComponent(accountId)}/detach-rule/${presetId}`,
    { method: 'POST' },
  );
}

/**
 * A preset carries every field the write endpoint requires, so an edit that
 * changes one attribute round-trips the rest unchanged.
 */
export function presetToWriteRequest(
  preset: RulePresetPayload,
  overrides: Partial<RulePresetWriteRequest> = {},
): RulePresetWriteRequest {
  return {
    name: preset.name,
    action: preset.action,
    level: preset.level,
    enabled: preset.enabled,
    conditions: preset.conditions,
    condition_logic: preset.condition_logic,
    cooldown_minutes: preset.cooldown_minutes,
    check_interval_minutes: preset.check_interval_minutes,
    notify_tg: preset.notify_tg,
    budget_change_percent: preset.budget_change_percent,
    budget_max_daily: preset.budget_max_daily,
    ...overrides,
  };
}
